bftengine/src/bftengine/PersistentStorageImp.cpp

```diff
@@ -0,0 +1,905 @@
+  // TODO: Yulia Serialize/deserialize booleans as uint8_t
+  const size_t sizeofBoolean = sizeof(boolean);
+  char buf[sizeofBoolean];
+  char *movablePtr = buf;
+  SeqNumData::serializeBoolean(movablePtr, boolean);
+  metadataStorage_->writeInTransaction(
+      SEQ_NUM_WINDOW + parameterId + SeqNumWindow::convertIndex(seqNum, seqNumWindowBeginning_), buf, sizeofBoolean);
+}
+
+void PersistentStorageImp::setForceCompletedInSeqNumWindow(const SeqNum seqNum, const bool forceCompleted) {
+  setBooleanInSeqNumWindow(seqNum, FORCE_COMPLETED, forceCompleted);
+}
+
+void PersistentStorageImp::setSlowStartedInSeqNumWindow(const SeqNum seqNum, const bool slowStarted) {
+  setBooleanInSeqNumWindow(seqNum, SLOW_STARTED, slowStarted);
+}
+
+void PersistentStorageImp::setCompletedMarkInCheckWindow(const SeqNum seqNum, const bool completed) {
+  Assert(completed);
+  const size_t sizeOfCompleted = sizeof(completed);
+  char buf[sizeOfCompleted];
+  char *movablePtr = buf;
+  CheckData::serializeCompletedMark(movablePtr, completed);
+  metadataStorage_->writeInTransaction(
+      CHECK_WINDOW + COMPLETED_MARK + CheckWindow::convertIndex(seqNum, checkWindowBeginning_), buf, sizeOfCompleted);
+}
+
+void PersistentStorageImp::setCheckpointMsgInCheckWindow(const SeqNum seqNum, const CheckpointMsg *const msg) {
+  size_t bufLen = CheckData::maxCheckpointMsgSize();
+  char *buf = new char[bufLen];
+  char *movablePtr = buf;
+  size_t actualSize = CheckData::serializeCheckpointMsg(movablePtr, (CheckpointMsg *) msg);
+  Assert(actualSize != 0);
+  metadataStorage_->writeInTransaction(
+      CHECK_WINDOW + CHECKPOINT_MSG + CheckWindow::convertIndex(seqNum, checkWindowBeginning_), buf, actualSize);
+  delete[] buf;
+}
+
+/***** Getters *****/
+
+string PersistentStorageImp::getStoredVersion() {
+  Assert(!isInWriteTran());
+  uint32_t outActualObjectSize = 0;
+  char *outBuf = new char[maxVersionSize_];
+  char *outBufPtr = outBuf;
+  metadataStorage_->read(VERSION_PARAMETER, maxVersionSize_, outBufPtr, outActualObjectSize);
+  if (!outActualObjectSize) // Parameter not found
+    return string();
+
+  uint32_t sizeOfVersion = 0;
+  memcpy(&sizeOfVersion, outBufPtr, sizeof(sizeOfVersion));
+  outBufPtr += sizeof(sizeOfVersion);
+  string savedVersion;
+  savedVersion.assign(outBufPtr, sizeOfVersion);
+  Assert(version_ == savedVersion);
+
+  delete[] outBuf;
+  return version_;
+}
+
+ReplicaConfig PersistentStorageImp::getReplicaConfig() {
+  uint32_t outActualObjectSize = 0;
+  UniquePtrToChar outBuf(new char[ReplicaConfigSerializer::maxSize(numOfReplicas_)]);
+  metadataStorage_->read(REPLICA_CONFIG, ReplicaConfigSerializer::maxSize(numOfReplicas_),
+                         outBuf.get(), outActualObjectSize);
+  SharedPtrToClass replicaConfigPtr = ReplicaConfigSerializer::deserialize(outBuf, outActualObjectSize);
+  auto *configPtr = ((ReplicaConfigSerializer *) (replicaConfigPtr.get()))->getConfig();
```

Comment:
Why is this cast needed?

Reply:
*Serializable -> *ReplicaConfigSerializer

Reply:
The base class (Serializable) does not support getConfig function.